File: app/Sources/DynoKit/Sampler.swift
import Foundation

/// Composes the individual collectors into one `Snapshot` per interval.
///
/// Marked `@unchecked Sendable` deliberately, and only two entry points may be
/// used concurrently: `sample()` must stay on one queue, while
/// `refreshModels(processes:)` touches nothing but the lock-protected
/// `ModelMonitor` and is safe to call from anywhere.
public final class Sampler: @unchecked Sendable {
    public let system: SystemInfo

    private let gpuStatsGroups: [IOReportSubscription.Group] = [
        .init("GPU Stats", "GPU Performance States"),
        .init("GPU Stats", "GPU Throttler Counters"),
        .init("GPU Stats", "PPM Target as % of Max GPU Power"),
    ]
    /// The memory-controller bandwidth histograms live under a per-die power
    /// management group. PMP0 is the usual name; alternates are tried only when
    /// it yields nothing, so the common case pays for one lookup.
    private let bandwidthGroup = IOReportSubscription.Group("PMP0", "DCS BW")
    private let bandwidthFallbacks = [
        IOReportSubscription.Group("PMP", "DCS BW"),
        IOReportSubscription.Group("PMP1", "DCS BW"),
    ]

    private static let energyChannels: Set<String> = [
        "GPU Energy", "GPU", "CPU Energy", "DRAM", "ANE",
    ]
    private static let bandwidthChannels: Set<String> = ["AMCC RD", "AMCC WR"]

    /// Above this share of the Metal working-set budget, further allocations
    /// start competing with the rest of the system.
    private static let gpuMemoryWarnPercent = 90.0
    private static let compressorWarnBytes = Int64(8 * GB)
    /// Scanning processes costs more than every other collector combined, and
    /// process memory does not move on a sub-second timescale.
    private static let processRefreshInterval: TimeInterval = 2.0

    private var subscription: IOReportSubscription?
    private let cpuLoad = CPULoadTracker()
    private let processScanner = ProcessScanner()
    private var previousSwapouts: UInt64?
    private var cachedProcesses: [ProcessSample] = []
    private var processesRefreshedAt: CFAbsoluteTime = 0
    private let modelMonitor = ModelMonitor()

    /// Below this the GPU is not really generating, so a bandwidth-derived
    /// rate would be noise rather than throughput. Idle sits well above zero on
    /// a Mac that is merely drawing its own desktop.
    private static let generatingGPUPercent = 25.0
    private static let idleGPUPercent = 15.0

    public var processLimit = 8
    public var processMinimumMemory = Int64(GB)

    public init(peakBandwidthOverride: Double? = nil) {
        var info = MachineInfo.gather()
        if let peakBandwidthOverride { info.peakBandwidthGBps = peakBandwidthOverride }
        system = info

        // Prime the process scanner so the first panel shows real names and
        // memory immediately; CPU shares fill in on the next scan.
        cachedProcesses = processScanner.scan(
            gpuPIDs: GPURegistry.clientPIDs(), minimumMemory: processMinimumMemory, limit: 8
        )
        processesRefreshedAt = CFAbsoluteTimeGetCurrent()

        subscription = Self.subscribe(gpuStatsGroups + [.init("Energy Model"), bandwidthGroup])
        if let current = subscription, !current.hasSubgroup("DCS BW") {
            // Some SoCs name the power-management group differently; pay for
            // the extra lookups only on those.
            current.close()
            subscription = Self.subscribe(
                gpuStatsGroups + [.init("Energy Model"), bandwidthGroup] + bandwidthFallbacks
            )
        }
    }

    private static func subscribe(
        _ groups: [IOReportSubscription.Group]
    ) -> IOReportSubscription? {
        if let subscription = try? IOReportSubscription(groups: groups) { return subscription }
        // Fall back to the whole GPU Stats group on SoCs that name their
        // performance-state subgroup differently.
        return try? IOReportSubscription(groups: [.init("GPU Stats"), .init("Energy Model")])
    }

    public var isAvailable: Bool { subscription != nil }

    public func close() {
        subscription?.close()
        subscription = nil
    }

    private static func keepChannel(_ group: String, _ subgroup: String?, _ name: String) -> Bool {
        if group == "Energy Model" { return energyChannels.contains(name) }
        if subgroup == "DCS BW" { return bandwidthChannels.contains(name) }
        if group == "GPU Stats" {
            if subgroup == "GPU Throttler Counters" { return name == "Throttle Counter Total" }
            return true
        }
        return false
    }

    public func sample() -> Snapshot {
        var snapshot = Snapshot()
        guard let subscription else { return snapshot }

        let (channels, interval) = subscription.sample(keep: Self.keepChannel)
        snapshot.interval = interval

        var bandwidthClipped = false
        for channel in channels {
            if isGPUResidency(channel) {
                let decoded = decodeResidency(channel)
                snapshot.gpu.busyPercent = decoded.busy
                snapshot.gpu.frequencyMHz = decoded.megahertz
                snapshot.gpu.stateResidency = decoded.residency
            } else if channel.group == "GPU Stats", channel.name == "Throttle Counter Total" {
                snapshot.gpu.throttleEvents = channel.value
            } else if channel.group == "GPU Stats", channel.format == .state,
                      channel.subgroup?.hasPrefix("PPM Target") == true {
                snapshot.gpu.powerTargetPercent = Self.percentMean(channel.states)
            } else if channel.group == "Energy Model", let joules = channel.joules {
                let watts = joules / interval
                switch channel.name {
                case "GPU Energy": snapshot.power.gpuWatts = watts
                case "GPU": if snapshot.power.gpuWatts == nil { snapshot.power.gpuWatts = watts }
                case "CPU Energy": snapshot.power.cpuWatts = watts
                case "DRAM": snapshot.power.dramWatts = watts
                case "ANE": snapshot.power.aneWatts = watts
                default: break
                }
            } else if channel.subgroup == "DCS BW",
                      let result = Self.histogramMean(channel.states) {
                bandwidthClipped = bandwidthClipped || result.clipped
                if channel.name == "AMCC RD" {
                    snapshot.bandwidth.readGBps = result.mean
                } else {
                    snapshot.bandwidth.writeGBps = result.mean
                }
            }
        }

        let accelerator = GPURegistry.accelerator()
        snapshot.gpu.allocatedBytes = accelerator.allocatedBytes
        snapshot.gpu.inUseBytes = accelerator.inUseBytes

        let vm = VirtualMemory.statistics()
        let swap = Sysctl.swapUsage()
        snapshot.memory.total = system.totalMemory
        snapshot.memory.app = vm.app
        snapshot.memory.wired = vm.wired
        snapshot.memory.compressed = vm.compressed
        snapshot.memory.free = vm.free
        snapshot.memory.used = vm.app + vm.wired + vm.compressed
        snapshot.memory.swapUsed = swap.used
        snapshot.memory.swapTotal = swap.total
        snapshot.memory.gpuBudget = system.gpuMemoryBudget
        snapshot.memory.gpuUsed = accelerator.inUseBytes

        var power = PowerSourceReader.read()
        power.gpuWatts = snapshot.power.gpuWatts
        power.cpuWatts = snapshot.power.cpuWatts
        power.dramWatts = snapshot.power.dramWatts
        power.aneWatts = snapshot.power.aneWatts
        snapshot.power = power

        snapshot.cpuPercent = cpuLoad.sample()
        snapshot.processes = refreshedProcesses()
        snapshot.models = tokenRates(
            for: modelMonitor.current,
            readGBps: snapshot.bandwidth.readGBps,
            gpuBusy: snapshot.gpu.busyPercent
        )
        snapshot.warnings = warnings(for: snapshot, swapouts: vm.swapouts, clipped: bandwidthClipped)
        return snapshot
    }

    /// Re-probe local model servers. Runs on its own cadence because it does
    /// loopback HTTP, unlike the counter reads in `sample()`.
    ///
    /// The process list is passed in rather than read from the cache: this runs
    /// on a different task from `sample()`, and the cache is not synchronised.
    public func refreshModels(processes: [ProcessSample]) async {
        await modelMonitor.refresh(processes: processes)
    }

    /// Fill in a token rate for every model that did not report one itself.
    ///
    /// During decode a dense model re-reads its entire weight set once per
    /// step, so `read bandwidth / weight bytes` is the step rate — which for a
    /// single stream is tokens per second. Measured against a 27B 8-bit model
    /// on an M5 Max this lands within a few percent. It only holds when one
    /// model owns the GPU, so anything else reports no rate rather than a
    /// number that looks authoritative and is not.
    private func tokenRates(
        for models: [LLMModel], readGBps: Double?, gpuBusy: Double
    ) -> [LLMModel] {
        let needingEstimate = models.filter { $0.rateSource != .measured }
        let attributable = needingEstimate.count == 1 && models.count == 1

        return models.map { model in
            guard model.rateSource != .measured else { return model }
            var model = model

            // A server that reports its own request count tells us directly
            // whether it is generating; no need to infer it from GPU load.
            if let stats = model.stats {
                model.tokensPerSecond = stats.activeRequests == 0 ? 0 : nil
                model.rateSource = stats.activeRequests == 0 ? .idle : .unavailable
                return model
            }

            if gpuBusy < Self.idleGPUPercent {
                model.tokensPerSecond = 0
                model.rateSource = .idle
            } else if attributable,
                      gpuBusy >= Self.generatingGPUPercent,
                      let readGBps, readGBps > 0,
                      let sizeGB = model.sizeGB, sizeGB > 0.5 {
                model.tokensPerSecond = readGBps / sizeGB
                model.rateSource = .estimated
            } else {
                model.tokensPerSecond = nil
                model.rateSource = .unavailable
            }
            return model
        }
    }

    private func refreshedProcesses() -> [ProcessSample] {
        let now = CFAbsoluteTimeGetCurrent()
        if !cachedProcesses.isEmpty,
           now - processesRefreshedAt < Self.processRefreshInterval {
            return cachedProcesses
        }
        cachedProcesses = processScanner.scan(
            gpuPIDs: GPURegistry.clientPIDs(),
            minimumMemory: processMinimumMemory,
            limit: processLimit
        )
        processesRefreshedAt = now
        return cachedProcesses
    }

    private func isGPUResidency(_ channel: IOReportChannel) -> Bool {
        guard channel.format == .state, channel.group == "GPU Stats" else { return false }
        let names = Set(channel.states.map(\.name))
        return names.contains("OFF") && names.contains("P1")
    }

    /// GPU busy share and average clock, from time spent per P-state.
    private func decodeResidency(
        _ channel: IOReportChannel
    ) -> (busy: Double, megahertz: Double?, residency: [String: Double]) {
        let total = channel.states.reduce(Int64(0)) { $0 + $1.residency }
        guard total > 0 else { return (0, nil, [:]) }

        var residency: [String: Double] = [:]
        for state in channel.states {
            residency[state.name] = 100 * Double(state.residency) / Double(total)
        }
        let busy = max(0, 100 - (residency["OFF"] ?? 0))

        var megahertz: Double?
        let frequencies = system.gpuFrequenciesMHz
        if !frequencies.isEmpty {
            var weighted = 0.0
            var activeShare = 0.0
            for (index, state) in channel.states.enumerated() {
                guard state.name != "OFF", index < frequencies.count else { continue }
                let share = Double(state.residency) / Double(total)
                weighted += frequencies[index] * share
                activeShare += share
            }
            // The clock the GPU runs at while busy, not averaged with idle.
            megahertz = activeShare > 0 ? weighted / activeShare : nil
        }
        return (busy, megahertz, residency)
    }

    /// Mean of a bandwidth histogram whose state names are bucket upper edges.
    /// `clipped` marks residency in the top bucket: the real value may be higher.
    static func histogramMean(
        _ states: [(name: String, residency: Int64)]
    ) -> (mean: Double, clipped: Bool)? {
        var edges: [(edge: Double, residency: Int64)] = []
        edges.reserveCapacity(states.count)
        for state in states {
            guard let edge = parseBandwidth(state.name) else { return nil }
            edges.append((edge, state.residency))
        }
        guard edges.count >= 2 else { return nil }
        let total = edges.reduce(Int64(0)) { $0 + $1.residency }
        guard total > 0 else { return nil }
        let width = edges[1].edge - edges[0].edge
        let mean = edges.reduce(0.0) {
            $0 + ($1.edge - width / 2) * Double($1.residency)
        } / Double(total)
        return (mean, edges.last!.residency > 0)
    }

    private static func parseBandwidth(_ label: String) -> Double? {
        let trimmed = label.trimmingCharacters(in: .whitespaces)
        let scales: [(String, Double)] = [("GB/s", 1), ("MB/s", 1.0 / 1024), ("KB/s", 1.0 / (1024 * 1024))]
        for (suffix, scale) in scales where trimmed.hasSuffix(suffix) {
            let number = trimmed.dropLast(suffix.count).trimmingCharacters(in: .whitespaces)
            guard let value = Double(number) else { return nil }
            return value * scale
        }
        return nil
    }

    private static func percentMean(_ states: [(name: String, residency: Int64)]) -> Double? {
        var values: [(value: Double, residency: Int64)] = []
        for state in states {
            let trimmed = state.name.trimmingCharacters(in: .whitespaces)
            guard trimmed.hasSuffix("%"),
                  let value = Double(trimmed.dropLast()) else { return nil }
            values.append((value, state.residency))
        }
        let total = values.reduce(Int64(0)) { $0 + $1.residency }
        guard total > 0 else { return nil }
        return values.reduce(0.0) { $0 + $1.value * Double($1.residency) } / Double(total)
    }

    private func warnings(
        for snapshot: Snapshot, swapouts: UInt64, clipped: Bool
    ) -> [String] {
        var messages: [String] = []

        if let usedPercent = snapshot.memory.gpuUsedPercent,
           usedPercent >= Self.gpuMemoryWarnPercent {
            messages.append(String(
                format: "GPU memory at %.0f%% of the Metal working-set budget — further allocations may spill to CPU memory",
                usedPercent
            ))
        }

        if let previous = previousSwapouts, swapouts > previous {
            let written = Double(swapouts - previous) * Double(VirtualMemory.pageSize)
            messages.append(
                "Actively swapping (\(Format.bytes(written)) since last sample) — token throughput will drop sharply"
            )
        }
        previousSwapouts = swapouts

        if snapshot.memory.compressed >= Self.compressorWarnBytes {
            messages.append(
                "Memory compressor holding \(Format.bytes(snapshot.memory.compressed)) — the model no longer fits comfortably"
            )
        }

        if snapshot.gpu.throttleEvents > 0 {
            messages.append("GPU throttled \(snapshot.gpu.throttleEvents) time(s) this interval")
        }

        if let system = snapshot.power.systemWatts,
           let adapter = snapshot.power.adapterMaxWatts, system > adapter {
            messages.append(String(
                format: "Drawing %.0f W from a %.0f W adapter — the battery is making up the difference",
                system, adapter
            ))
        } else if snapshot.power.onBattery {
            messages.append("On battery — the GPU may be held below its plugged-in clocks")
        }

        if clipped {
            messages.append(
                "Memory bandwidth is saturating the controller's top histogram bucket — the real figure may be higher than shown"
            )
        }
        return messages
    }
}

File: src/dyno/monitor/sampler.py
"""Composes the individual collectors into a single Snapshot per interval."""

from __future__ import annotations

import re
import time
from typing import Iterable

from . import sources
from .cf import metal_device_info
from .ioreport import FORMAT_STATE, Channel, IOReportError, Subscription
from .model import (
    BandwidthSample,
    GB,
    GpuSample,
    MemorySample,
    PowerSample,
    ProcessSample,
    Snapshot,
    SystemInfo,
)

_GPU_STATS_SUBGROUPS = (
    ("GPU Stats", "GPU Performance States"),
    ("GPU Stats", "GPU Throttler Counters"),
    ("GPU Stats", "PPM Target as % of Max GPU Power"),
)
# The memory-controller bandwidth histograms live under a per-die power
# management group. PMP0 is the usual name; the alternates are tried only if it
# yields nothing, so the common case pays for one lookup.
_BANDWIDTH_GROUP = ("PMP0", "DCS BW")
_BANDWIDTH_FALLBACKS = (("PMP", "DCS BW"), ("PMP1", "DCS BW"))

_BUCKET_LABEL = re.compile(r"([0-9.]+)\s*(GB|MB|KB)/s")
_UNIT_TO_GB = {"GB": 1.0, "MB": 1 / 1024, "KB": 1 / 1024 ** 2}
_PERCENT_LABEL = re.compile(r"([0-9.]+)\s*%")

# Channels actually read out of each group. Everything else in the subscribed
# subgroups is skipped before its state array is decoded.
_ENERGY_CHANNELS = frozenset({"GPU Energy", "GPU", "CPU Energy", "DRAM", "ANE"})
_BANDWIDTH_CHANNELS = frozenset({"AMCC RD", "AMCC WR"})

# Scanning the process table costs far more than every other collector
# combined, and process memory does not move on a 100 ms timescale.
_PROCESS_REFRESH_SECONDS = 2.0

# Above this share of the Metal working-set budget, further allocations start
# competing with the rest of the system.
_GPU_MEMORY_WARN_PCT = 90.0
_COMPRESSOR_WARN = 8 * GB


def _is_gpu_residency(channel: Channel) -> bool:
    if channel.format != FORMAT_STATE or channel.group != "GPU Stats":
        return False
    names = {name for name, _ in channel.states}
    return "OFF" in names and "P1" in names


def _histogram_mean(states: Iterable[tuple[str, int]]) -> tuple[float, bool] | None:
    """Mean of a bandwidth histogram whose state names are bucket upper edges.

    Returns (GB/s, clipped) where ``clipped`` marks residency in the top bucket,
    meaning the real value may be higher than reported.
    """
    edges: list[tuple[float, int]] = []
    for label, residency in states:
        match = _BUCKET_LABEL.search(label)
        if not match:
            return None
        edges.append((float(match.group(1)) * _UNIT_TO_GB[match.group(2)], residency))
    total = sum(residency for _, residency in edges)
    if total <= 0 or len(edges) < 2:
        return None
    width = edges[1][0] - edges[0][0]
    mean = sum((edge - width / 2) * residency for edge, residency in edges) / total
    return mean, edges[-1][1] > 0


def _state_percent_mean(states: Iterable[tuple[str, int]]) -> float | None:
    values: list[tuple[float, int]] = []
    for label, residency in states:
        match = _PERCENT_LABEL.search(label)
        if not match:
            return None
        values.append((float(match.group(1)), residency))
    total = sum(residency for _, residency in values)
    if total <= 0:
        return None
    return sum(value * residency for value, residency in values) / total


class Sampler:
    """Holds the IOReport subscription and turns each interval into a Snapshot."""

    def __init__(self, *, peak_bandwidth: float | None = None, process_limit: int = 8,
                 process_min_memory: int = GB) -> None:
        self.process_limit = process_limit
        self.process_min_memory = process_min_memory
        self._cpu_load = sources.CpuLoad()
        self._previous_swapouts: int | None = None
        self._processes: list[ProcessSample] = []
        self._processes_at = 0.0

        accelerator = sources.accelerator_properties()
        metal = metal_device_info()
        chip = (
            metal.get("name")
            or accelerator.get("model")
            or sources.sysctl_str("machdep.cpu.brand_string")
            or "Apple Silicon"
        )
        gpu_cores = accelerator.get("cores")
        total_memory = sources.sysctl_int("hw.memsize") or 0

        self.system = SystemInfo(
            chip=chip,
            gpu_cores=gpu_cores if isinstance(gpu_cores, int) else None,
            cpu_cores=sources.sysctl_int("hw.ncpu") or 0,
            performance_cores=sources.sysctl_int("hw.perflevel0.logicalcpu"),
            efficiency_cores=sources.sysctl_int("hw.perflevel1.logicalcpu"),
            total_memory=total_memory,
            gpu_memory_budget=metal.get("recommended_max_working_set"),
            gpu_frequencies_mhz=sources.gpu_frequency_table(),
            peak_memory_bandwidth_gbps=peak_bandwidth
            or sources.peak_bandwidth_gbps(chip, gpu_cores if isinstance(gpu_cores, int) else None),
            macos_version=sources.macos_version(),
        )

        self._subscription = self._subscribe((_BANDWIDTH_GROUP,))
        if not self._subscription.has_subgroup("DCS BW"):
            # Some SoCs name the power-management group differently; pay for the
            # extra lookups only on those.
            self._subscription.close()
            self._subscription = self._subscribe(
                (_BANDWIDTH_GROUP,) + _BANDWIDTH_FALLBACKS
            )

    def _subscribe(self, bandwidth_groups: tuple[tuple[str, str | None], ...]) -> Subscription:
        groups = list(_GPU_STATS_SUBGROUPS) + [("Energy Model", None)] + list(bandwidth_groups)
        try:
            return Subscription(groups)
        except IOReportError:
            # Fall back to the whole GPU Stats group on SoCs that name their
            # performance-state subgroup differently.
            return Subscription([("GPU Stats", None), ("Energy Model", None)])



    def close(self) -> None:
        self._subscription.close()

    @staticmethod
    def _keep_channel(group: str, subgroup: str | None, name: str) -> bool:
        if group == "Energy Model":
            return name in _ENERGY_CHANNELS
        if subgroup == "DCS BW":
            return name in _BANDWIDTH_CHANNELS
        if group == "GPU Stats":
            if subgroup == "GPU Throttler Counters":
                return name == "Throttle Counter Total"
            return True
        return False

    def _process_samples(self) -> list[ProcessSample]:
        now = time.monotonic()
        if self._processes and now - self._processes_at < _PROCESS_REFRESH_SECONDS:
            return self._processes
        gpu_pids = sources.gpu_client_pids()
        self._processes = [
            ProcessSample(**row)
            for row in sources.process_table(
                gpu_pids, self.process_min_memory, self.process_limit
            )
        ]
        self._processes_at = now
        return self._processes

    def sample(self) -> Snapshot:
        channels, interval = self._subscription.sample(self._keep_channel)
        interval = max(interval, 1e-6)

        gpu_active = 0.0
        gpu_mhz: float | None = None
        residency: dict[str, float] = {}
        throttle_events = 0
        power_target: float | None = None
        gpu_w = cpu_w = dram_w = ane_w = None
        read_gbps = write_gbps = None
        bandwidth_clipped = False

        for channel in channels:
            if _is_gpu_residency(channel):
                gpu_active, gpu_mhz, residency = self._decode_residency(channel)
            elif channel.group == "GPU Stats" and channel.name == "Throttle Counter Total":
                throttle_events = channel.value or 0
            elif channel.group == "GPU Stats" and channel.format == FORMAT_STATE and (
                channel.subgroup or ""
            ).startswith("PPM Target"):
                power_target = _state_percent_mean(channel.states)
            elif channel.group == "Energy Model":
                joules = channel.energy_joules()
                if joules is None:
                    continue
                watts = joules / interval
                if channel.name == "GPU Energy":
                    gpu_w = watts
                elif channel.name == "GPU" and gpu_w is None:
                    gpu_w = watts
                elif channel.name == "CPU Energy":
                    cpu_w = watts
                elif channel.name == "DRAM":
                    dram_w = watts
                elif channel.name == "ANE":
                    ane_w = watts
            elif channel.subgroup == "DCS BW" and channel.name in ("AMCC RD", "AMCC WR"):
                result = _histogram_mean(channel.states)
                if result is None:
                    continue
                mean, clipped = result
                bandwidth_clipped = bandwidth_clipped or clipped
                if channel.name == "AMCC RD":
                    read_gbps = mean
                else:
                    write_gbps = mean

        accelerator = sources.accelerator_properties()
        stats = accelerator.get("stats") or {}
        gpu_allocated = stats.get("Alloc system memory")
        gpu_in_use = stats.get("In use system memory")

        vm = sources.vm_statistics()
        swap_used, swap_total = sources.swap_usage()
        total_memory = self.system.total_memory
        wired = vm.get("wired", 0)
        compressed = vm.get("compressed", 0)
        app = vm.get("app", 0)
        used = app + wired + compressed
        pressure = 100.0 * (wired + compressed) / total_memory if total_memory else 0.0

        memory = MemorySample(
            total=total_memory,
            used=used,
            app=app,
            wired=wired,
            compressed=compressed,
            free=vm.get("free", 0),
            cached=vm.get("external", 0),
            swap_used=swap_used,
            swap_total=swap_total,
            pressure_pct=pressure,
            gpu_budget=self.system.gpu_memory_budget,
            gpu_used=gpu_in_use if isinstance(gpu_in_use, int) else None,
        )

        source = sources.power_source()
        power = PowerSample(
            gpu_w=gpu_w,
            cpu_w=cpu_w,
            dram_w=dram_w,
            ane_w=ane_w,
            system_w=source["system_w"],
            adapter_max_w=source["adapter_max_w"],
            on_battery=source["on_battery"],
            battery_pct=source["battery_pct"],
        )

        processes = self._process_samples()

        snapshot = Snapshot(
            timestamp=time.time(),
            interval=interval,
            gpu=GpuSample(
                active_pct=gpu_active,
                frequency_mhz=gpu_mhz,
                pstate_residency=residency,
                allocated_bytes=gpu_allocated if isinstance(gpu_allocated, int) else None,
                in_use_bytes=gpu_in_use if isinstance(gpu_in_use, int) else None,
                throttle_events=throttle_events,
                power_target_pct=power_target,
            ),
            memory=memory,
            power=power,
            bandwidth=BandwidthSample(
                dram_read_gbps=read_gbps, dram_write_gbps=write_gbps, gpu_gbps=None
            ),
            cpu_pct=self._cpu_load.sample(),
            processes=processes,
        )
        snapshot.warnings = self._warnings(snapshot, vm)
        if bandwidth_clipped:
            snapshot.warnings.append(
                "Memory bandwidth is saturating the controller's top histogram "
                "bucket - the real figure may be higher than shown"
            )
        return snapshot

    def _decode_residency(self, channel: Channel) -> tuple[float, float | None, dict[str, float]]:
        """GPU busy percentage and average clock, from time spent per P-state."""
        total = sum(value for _, value in channel.states)
        if total <= 0:
            return 0.0, None, {}
        residency = {name: 100.0 * value / total for name, value in channel.states}
        idle = residency.get("OFF", 0.0)
        active = max(0.0, 100.0 - idle)

        frequencies = self.system.gpu_frequencies_mhz
        mhz: float | None = None
        if frequencies:
            weighted = 0.0
            active_share = 0.0
            for index, (name, value) in enumerate(channel.states):
                if name == "OFF" or index >= len(frequencies):
                    continue
                share = value / total
                weighted += frequencies[index] * share
                active_share += share
            # Report the clock the GPU runs at while busy, not averaged with idle.
            mhz = weighted / active_share if active_share > 0 else None
        return active, mhz, residency

    def _warnings(self, snapshot: Snapshot, vm: dict[str, int]) -> list[str]:
        warnings: list[str] = []
        memory = snapshot.memory

        used_pct = memory.gpu_used_pct
        if used_pct is not None and used_pct >= _GPU_MEMORY_WARN_PCT:
            warnings.append(
                "GPU memory at %.0f%% of the Metal working-set budget - "
                "further allocations may spill to CPU memory" % used_pct
            )

        swapouts = vm.get("swapouts", 0)
        if self._previous_swapouts is not None and swapouts > self._previous_swapouts:
            warnings.append(
                "Actively swapping (%s written since last sample) - "
                "token throughput will drop sharply"
                % _format_bytes((swapouts - self._previous_swapouts) * sources.PAGE_SIZE)
            )
        self._previous_swapouts = swapouts

        if memory.compressed >= _COMPRESSOR_WARN:
            warnings.append(
                "Memory compressor holding %s - the model no longer fits comfortably"
                % _format_bytes(memory.compressed)
            )

        if snapshot.gpu.throttle_events > 0:
            warnings.append(
                "GPU throttled %d time(s) this interval" % snapshot.gpu.throttle_events
            )

        power = snapshot.power
        if (
            power.system_w is not None
            and power.adapter_max_w
            and power.system_w > power.adapter_max_w
        ):
            warnings.append(
                "Drawing %.0f W from a %.0f W adapter - the battery is making up the difference"
                % (power.system_w, power.adapter_max_w)
            )
        elif power.on_battery:
            warnings.append("On battery - the GPU may be held below its plugged-in clocks")

        return warnings


def _format_bytes(value: float) -> str:
    for unit, scale in (("TB", 1024 ** 4), ("GB", GB), ("MB", 1024 ** 2), ("KB", 1024)):
        if abs(value) >= scale:
            return "%.1f %s" % (value / scale, unit)
    return "%d B" % value
